OneSignalSDK/onesignal/src/main/java/com/onesignal/OneSignalCacheCleaner.java

```diff
@@ -4,69 +4,188 @@
+                String[] whereArgs = new String[]{sixMonthsAgoInSeconds};
+
+                Cursor cursor = writableDb.query(OneSignalDbContract.InAppMessageTable.TABLE_NAME,
+                        retColumns,
+                        whereStr,
+                        whereArgs,
+                        null,
+                        null,
+                        null);
+
+                // From cursor get all of the old message ids and old clicked click ids
+                Set<String> oldMessageIds = OSUtils.newConcurrentSet();
+                Set<String> oldClickedClickIds = OSUtils.newConcurrentSet();
+                if (cursor.moveToFirst()) {
+                    do {
+                        String oldMessageId = cursor.getString(
+                                cursor.getColumnIndex(
+                                        OneSignalDbContract.InAppMessageTable.COLUMN_NAME_MESSAGE_ID));
+                        String oldClickIds = cursor.getString(
+                                cursor.getColumnIndex(
+                                        OneSignalDbContract.InAppMessageTable.COLUMN_NAME_MESSAGE_ID));
+
+                        oldMessageIds.add(oldMessageId);
+                        oldClickedClickIds.addAll(OSUtils.newStringSetFromString(oldClickIds));
+                    } while (cursor.moveToNext());
+                }
+                cursor.close();
+
+                // 2. Delete old IAMs from SQL
+                writableDb.delete(
+                        OneSignalDbContract.InAppMessageTable.TABLE_NAME,
+                        whereStr,
+                        whereArgs);
+
+                // 3. Use queried data to clean SharedPreferences
+                cleanCachedSharedPreferenceIamData(oldMessageIds, oldClickedClickIds);
+            }
+
+            private void cleanCachedSharedPreferenceIamData(Set<String> oldMessageIds, Set<String> oldClickedClickIds) {
+                // IAMs without redisplay on with pile up and we need to clean these for dismissing, impressions, and clicks
+                Set<String> dismissedMessages = OneSignalPrefs.getStringSet(
+                        OneSignalPrefs.PREFS_ONESIGNAL,
+                        OneSignalPrefs.PREFS_OS_DISMISSED_IAMS,
+                        OSUtils.<String>newConcurrentSet());
+
+                Set<String> impressionedMessages = OneSignalPrefs.getStringSet(
+                        OneSignalPrefs.PREFS_ONESIGNAL,
+                        OneSignalPrefs.PREFS_OS_IMPRESSIONED_IAMS,
+                        OSUtils.<String>newConcurrentSet());
+
+                Set<String> clickedClickIds = OneSignalPrefs.getStringSet(
+                        OneSignalPrefs.PREFS_ONESIGNAL,
+                        OneSignalPrefs.PREFS_OS_CLICKED_CLICK_IDS_IAMS,
+                        OSUtils.<String>newConcurrentSet());
+
+                if (oldMessageIds != null) {
```

Comment:
small optimization, but you can avoid dismissedMessages and impressionedMessages shared preference query if you already know that oldMessageIds is null or empty